auth/default_authorizer.cc

```diff
@@ -307,6 +314,47 @@ future<> default_authorizer::revoke_all(const resource& resource) const {
     });
 }
 
+future<> default_authorizer::revoke_all(const resource& resource) const {
+    if (legacy_mode(_qp)) {
+        co_return co_await revoke_all_legacy(resource);
+    }
+    static const sstring query = format("SELECT {} FROM {}.{} WHERE {} = ? ALLOW FILTERING",
+        ROLE_NAME,
+        _auth_ks_name,
+        PERMISSIONS_CF,
+        RESOURCE_NAME);
+    try {
+        auto group0_guard = co_await _group0_client.start_operation();
+        auto timestamp = group0_guard.write_timestamp();
+
+        auto res = co_await _qp.execute_internal(
+            query,
+            db::consistency_level::LOCAL_ONE,
+            {resource.name()},
+            cql3::query_processor::cache_internal::no);
+
+        std::vector<mutation> muts;
+        co_await parallel_for_each(res->begin(), res->end(),
+            [this, timestamp, resource, &muts](const cql3::untyped_result_set::row& r) -> future<> {
+            static const sstring query = format("DELETE FROM {}.{} WHERE {} = ? AND {} = ?",
+                _auth_ks_name,
+                PERMISSIONS_CF,
+                ROLE_NAME,
+                RESOURCE_NAME);
+            auto muts2 = co_await _qp.get_mutations_internal(
+                query,
+                internal_distributed_query_state(),
+                timestamp,
+                {r.get_as<sstring>(ROLE_NAME), resource.name()});
+            muts.reserve(muts.size() + muts2.size());
```

Comment:
Doesn't this `reserve` make the loop quadratic? The whole parallel_for_each is `O(res.len() * total_number_of_mutations)`.

I think it's better to either remove the `reserve` or preallocate the correct capacity up front, if it is known.

Reply:
I think reserve make it "less" quadratic in terms of re-allocations. Complexity is indeed m*n but the number of deletions is not known. Total number of mutations depends on data, and I think performance will be similar if we do it in a concurrent loop or sequentially (with hope that current parallel_for_each is slightly faster).

Reply:
but I also worry here about muts size passed to prepare_command, perhaps I should batch those mutations and do announce_mutations_with_guard every X mutations collected (what's sensible X for raft?)

Reply:
> I think reserve make it "less" quadratic in terms of re-allocations. Complexity is indeed m*n but the number of deletions is not known. Total number of mutations depends on data, and I think performance will be similar if we do it in a concurrent loop or sequentially (with hope that current parallel_for_each is slightly faster).

If you only `push_back` elements to a vector without calling `reserve`, you get amortized `O(1)` insertion time because vector's storage will be reallocated every power or two or so. However, if you use reserve then I'm worried that the storage will be reallocated every time `reserve` is called.

See the note on the cppreference: https://en.cppreference.com/w/cpp/container/vector/reserve#:~:text=of%20the%20container.-,Notes,-Correctly%20using%20reserve

How many mutations does each `get_mutations_internal` call in `parallel_for_each` produce? I suspect it might be just one? If that's so then maybe we can just call `muts.reserve(res.size())` at the beginning and it will be good enough.

However, in the end it might be even better to use `utils::chunked_vector` which also avoids the large allocation problem.

Reply:
> but I also worry here about muts size passed to prepare_command, perhaps I should batch those mutations and do announce_mutations_with_guard every X mutations collected (what's sensible X for raft?)

There is a limit to the maximum allowed size of a mutation - it's a bit less than half of the commitlog segment size (by default `schema_commitlog_segment_size_in_mb == 128`)

We are already doing this kind of batching in the topology coordinator - you can take a look at `prepare_new_cdc_generation_data` in `storage_service.cc`. We obtain the maximum size of a raft command and split the data that we want to insert over multiple mutations.

Naturally, if you send it over multiple batches then you will lose atomicity. In case of CDC it's not a problem for us (I don't want to go into details), but here I'm not sure if it is.

Reply:
I think I'll have to do batching as I understand otherwise the function will break for some users (the ones having many roles or permissions).

Reply:
btw yes, you were right about reserve(), it defeats the 2*N capacity growth mechanism